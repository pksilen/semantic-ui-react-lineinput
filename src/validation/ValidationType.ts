export type PredefinedValidationType =
  | 'url'
  | 'emailAddress'
  | 'creditCardNumber'
  | 'creditCardExpiration'
  | 'creditCardVerificationCode'
  | 'number'
  | 'integer'
  | 'alphaNumeric'
  | 'usZipCode'
  | 'caPostCode'
  | 'ukPostCode'
  | 'phoneNumber'
  | 'usSSN'
  | 'ipAddress'
  | 'ipv4Address'
  | 'ipv6Address'
  | 'hexColor';

type ValidationFunction = (inputString: string) => boolean;

export type ValidationType = PredefinedValidationType | string[] | RegExp | ValidationFunction;
