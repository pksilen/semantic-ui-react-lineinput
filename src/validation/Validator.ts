// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
import is from 'is_js';
// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
import getBrowserLocale from 'browser-locale';
// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
import cardsy from 'cardsy';
import { parsePhoneNumberFromString } from 'libphonenumber-js/max';
import { ValidationType } from './ValidationType';

export default class Validator {
  // noinspection FunctionWithMoreThanThreeNegationsJS,OverlyComplexFunctionJS
  // eslint-disable-next-line consistent-return
  static validateValue(
    valueString: string,
    validation: ValidationType,
    minValue?: number,
    maxValue?: number,
    suppliedCountryCode?: string,
    minLength?: number
  ): boolean {
    if (minLength !== undefined && valueString.length < minLength)  {
      return false;
    }

    if (typeof validation === 'function') {
      return validation(valueString);
    }

    if (Array.isArray(validation)) {
      return validation.includes(valueString);
    }

    if (typeof validation === 'object') {
      return !!valueString.match(validation);
    }

    let valueAsNumber = 0;
    if (validation === 'number') {
      valueAsNumber = parseFloat(valueString);
    } else if (validation === 'integer') {
      if (valueString.includes('.')) {
        valueAsNumber = NaN;
      } else {
        valueAsNumber = parseInt(valueString, 10);
      }
    }

    const countryCode = suppliedCountryCode || getBrowserLocale().split('-')[1];

    switch (validation) {
      case 'url':
        return is.url(valueString);

      case 'emailAddress':
        return is.email(valueString);

      case 'creditCardNumber':
        return is.creditCard(valueString.replace(/\s/g, ''));

      case 'creditCardExpiration':
        return cardsy.validate.expiryString(valueString);

      case 'creditCardVerificationCode':
        return !!valueString.match(/^\d{3,4}$/);

      case 'number':
      case 'integer':
        return (
          !Number.isNaN(valueAsNumber) &&
          (minValue === undefined || valueAsNumber >= minValue) &&
          (maxValue === undefined || valueAsNumber <= maxValue)
        );

      case 'alphaNumeric':
        return is.alphaNumeric(valueString);

      case 'usZipCode':
        return is.usZipCode(valueString);

      case 'caPostCode':
        return is.caPostalCode(valueString);

      case 'ukPostCode':
        return is.ukPostCode(valueString);

      case 'phoneNumber':
        return !!parsePhoneNumberFromString(valueString, countryCode);

      case 'usSSN':
        return is.socialSecurityNumber(valueString);

      case 'ipAddress':
        return is.ip(valueString);

      case 'ipv4Address':
        return is.ipv4(valueString);

      case 'ipv6Address':
        return is.ipv6(valueString);

      case 'hexColor':
        return is.hexColor(valueString);

      // no default
    }
  }
}
