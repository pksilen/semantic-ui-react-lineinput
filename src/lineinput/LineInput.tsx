import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Icon, SemanticCOLORS, SemanticICONS } from 'semantic-ui-react';
// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
import cardsy from 'cardsy';
// eslint-disable-next-line import/no-named-default
import { default as getCreditCardType } from 'credit-card-type';
import { ValidationType } from '../validation/ValidationType';
import styleMap from '../styleMap';
import Validator from '../validation/Validator';

export interface Props extends React.HTMLAttributes<HTMLDivElement> {
  allowEmptyValue: boolean;
  className?: string;
  countryCode: string;
  creditCardNumber: string;
  disabled: boolean;
  errorText: string;
  errorTextPosition: 'bottom' | 'right';
  focus: boolean;
  icon: SemanticICONS;
  iconColor: SemanticCOLORS;
  iconPosition: 'left' | 'right';
  maxLength?: number;
  maxValue?: number;
  minLength?: number;
  minValue?: number;
  onValueChange: (value: string) => void;
  placeholder: string;
  size: 'mini' | 'small' | 'large' | 'big' | 'huge' | 'massive';
  type: string;
  validation?: ValidationType;
  validationSuccessIcon: SemanticICONS;
  validationErrorIcon: SemanticICONS;
  value: string;
}

type State = {
  hasValidValue: boolean;
  hasFocus: boolean;
  hasBeenFocused: boolean;
};

export default class LineInput extends React.Component<Props, {}> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  static propTypes: { [key in keyof Props]: any } = {
    allowEmptyValue: PropTypes.bool,
    className: PropTypes.string,
    countryCode: PropTypes.string,
    creditCardNumber: PropTypes.string,
    disabled: PropTypes.bool,
    icon: PropTypes.string,
    iconColor: PropTypes.string,
    iconPosition: PropTypes.oneOf(['left', 'right']),
    errorText: PropTypes.string,
    errorTextPosition: PropTypes.oneOf(['bottom', 'right']),
    focus: PropTypes.bool,
    maxLength: PropTypes.number,
    maxValue: PropTypes.number,
    minLength: PropTypes.number,
    minValue: PropTypes.number,
    onValueChange: PropTypes.func.isRequired,
    placeholder: PropTypes.string,
    size: PropTypes.oneOf(['mini', 'small', 'large', 'big', 'huge', 'massive']),
    type: PropTypes.string,
    validation: PropTypes.oneOfType([
      PropTypes.oneOf([
        'url',
        'emailAddress',
        'creditCardNumber',
        'creditCardExpiration',
        'creditCardVerificationCode',
        'number',
        'integer',
        'alphaNumeric',
        'usZipCode',
        'caPostCode',
        'ukPostCode',
        'phoneNumber',
        'usSSN',
        'ipAddress',
        'ipv4Address',
        'ipv6Address',
        'hexColor'
      ]),
      PropTypes.arrayOf(PropTypes.string),
      PropTypes.instanceOf(RegExp),
      PropTypes.func
    ]),
    validationErrorIcon: PropTypes.string,
    validationSuccessIcon: PropTypes.string,
    value: PropTypes.string.isRequired
  };

  static defaultProps = {
    allowEmptyValue: false,
    className: undefined,
    countryCode: '',
    creditCardNumber: '',
    disabled: false,
    errorText: '',
    errorTextPosition: 'bottom',
    focus: false,
    icon: '',
    iconColor: undefined,
    iconPosition: 'left',
    maxLength: undefined,
    maxValue: undefined,
    minLength: undefined,
    minValue: undefined,
    placeholder: '',
    size: 'small',
    type: undefined,
    validation: undefined,
    validationErrorIcon: '',
    validationSuccessIcon: ''
  };

  state: State = {
    hasValidValue: true,
    hasFocus: false,
    hasBeenFocused: false
  };

  onInputBlur = () => {
    const {
      allowEmptyValue,
      countryCode,
      creditCardNumber,
      maxValue,
      minLength,
      minValue,
      validation,
      value
    } = this.props;

    this.setState({
      hasFocus: false
    });

    if (allowEmptyValue && !value) {
      return;
    }

    if (validation) {
      const hasValidValue = Validator.validateValue(
        value,
        validation,
        minValue,
        maxValue,
        countryCode,
        minLength,
        creditCardNumber
      );

      this.setState({
        hasValidValue
      });
    }
  };

  onInputFocus = () => {
    this.setState({
      hasValidValue: true,
      hasFocus: true,
      hasBeenFocused: true
    });
  };

  onInputChange = ({ target: { value } }: React.ChangeEvent<HTMLInputElement>) => {
    const { onValueChange, validation } = this.props;

    if (validation === 'creditCardNumber') {
      onValueChange(cardsy.format.number(value));
    } else if (validation === 'creditCardExpiration') {
      const expiryRegExp = /^(\d{2})\/(\d{2})$/;
      const expiryMatches = value.match(expiryRegExp);
      const expiryStr = expiryMatches ? `${expiryMatches[1]} / ${expiryMatches[2]}` : value;
      onValueChange(cardsy.format.expiryString(expiryStr));
    } else if (validation === 'creditCardVerificationCode') {
      onValueChange(cardsy.format.cvc(value));
    } else {
      onValueChange(value);
    }
  };

  getInputType = (): string => {
    const { type, validation } = this.props;

    if (type) {
      return type;
    }

    switch (validation) {
      case 'emailAddress':
        return 'email';

      case 'phoneNumber':
        return 'tel';

      case 'url':
        return 'url';

      default:
        return 'text';
    }
  };

  // noinspection OverlyComplexFunctionJS
  getIconName = (): SemanticICONS => {
    const { icon, validation, validationErrorIcon, validationSuccessIcon, value } = this.props;
    const { hasFocus, hasBeenFocused, hasValidValue } = this.state;

    if (validationErrorIcon && !hasValidValue) {
      return hasFocus ? ('' as SemanticICONS) : validationErrorIcon;
    }

    if (validationSuccessIcon && hasValidValue) {
      return hasFocus || !hasBeenFocused ? ('' as SemanticICONS) : validationSuccessIcon;
    }

    if (validation === 'creditCardNumber' && value) {
      const creditCardTypes = getCreditCardType(value);
      const creditCardType = creditCardTypes && creditCardTypes[0] ? creditCardTypes[0].niceType : '';

      switch (creditCardType) {
        case 'Visa':
          return 'cc visa';

        case 'Mastercard':
          return 'cc mastercard';

        case 'American Express':
          return 'cc amex';

        case 'Diners Club':
          return 'cc diners club';

        case 'Discover':
          return 'cc discover';

        case 'JCB':
          return 'cc jcb';

        default:
          return 'credit card';
      }
    }

    return icon;
  };

  getIconColor = () => {
    const { iconColor, validationErrorIcon, validationSuccessIcon } = this.props;
    const { hasFocus, hasValidValue } = this.state;

    const hasValidationIcon = !hasFocus && (validationErrorIcon || validationSuccessIcon);

    if (hasValidationIcon && hasValidValue) {
      return 'green';
    }

    if (hasValidationIcon && !hasValidValue) {
      return 'red';
    }

    return iconColor;
  };

  hasIcon = () => {
    const { icon, validation, validationErrorIcon, validationSuccessIcon, value } = this.props;

    const hasValidationIcon = validationErrorIcon || validationSuccessIcon;
    const hasCreditCardIcon = validation === 'creditCardNumber' && value;

    return icon || hasValidationIcon || hasCreditCardIcon;
  };

  getClassName = () => {
    const { disabled, focus, iconPosition, size, validationErrorIcon, validationSuccessIcon } = this.props;

    let className = `ui input ${size}`;

    if (this.hasIcon()) {
      className = `${className} ${
        validationErrorIcon || validationSuccessIcon ? 'right' : iconPosition
      } icon`;
    }

    if (disabled) {
      className = `${className} disabled`;
    }

    if (focus) {
      className = `${className} focus`;
    }

    return className;
  };

  render(): React.ReactElement {
    const {
      allowEmptyValue,
      countryCode,
      creditCardNumber,
      disabled,
      errorText,
      errorTextPosition,
      focus,
      icon,
      iconColor,
      iconPosition,
      minLength,
      minValue,
      maxLength,
      maxValue,
      onValueChange,
      placeholder,
      size,
      style,
      type,
      validation,
      validationErrorIcon,
      validationSuccessIcon,
      value,
      ...restOfProps
    } = this.props;

    const { hasValidValue } = this.state;

    const topDivStyle =
      errorTextPosition === 'bottom'
        ? { ...styleMap.topDiv, ...styleMap.topDivWhenErrorAtBottom }
        : styleMap.topDiv;

    const inputStyle = hasValidValue ? styleMap.input : { ...styleMap.input, ...styleMap.errorInput };
    const labelStyle =
      errorTextPosition === 'bottom' ? styleMap.label : { ...styleMap.label, ...styleMap.rightLabel };
    const iconStyle = validationErrorIcon || validationSuccessIcon ? styleMap.validationIcon : {};

    // eslint-disable-next-line react/jsx-props-no-spreading
    return (
      <div style={topDivStyle} {...restOfProps}>
        <div className={this.getClassName()}>
          <input
            maxLength={maxLength}
            type={this.getInputType()}
            onBlur={this.onInputBlur}
            onFocus={this.onInputFocus}
            onChange={this.onInputChange}
            placeholder={placeholder}
            style={inputStyle}
            value={value}
          />
          {this.hasIcon() ? (
            <Icon color={this.getIconColor()} name={this.getIconName()} style={iconStyle} />
          ) : (
            undefined
          )}
        </div>
        {!hasValidValue && errorText ? (
          <div className={`ui label ${size}`} style={labelStyle}>
            {errorText}
          </div>
        ) : (
          undefined
        )}
      </div>
    );
  }
}
