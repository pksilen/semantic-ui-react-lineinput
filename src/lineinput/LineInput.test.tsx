import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Icon } from 'semantic-ui-react';
import { mount, shallow as renderShallow } from 'enzyme';
import LineInput from './LineInput';
import { ValidationType } from '../validation/ValidationType';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let onValueChangeMock: any;

beforeEach(() => {
  onValueChangeMock = jest.fn();
});

describe('propTypes', () => {
  it('should validate prop types', () => {
    // eslint-disable-next-line react/forbid-foreign-prop-types
    const { propTypes } = LineInput;

    expect(propTypes.allowEmptyValue).toBe(PropTypes.bool);
    expect(propTypes.className).toBe(PropTypes.string);
    expect(propTypes.countryCode).toBe(PropTypes.string);
    expect(propTypes.disabled).toBe(PropTypes.bool);
    expect(propTypes.errorText).toBe(PropTypes.string);
    expect(propTypes.errorTextPosition).toBeTruthy();
    expect(propTypes.icon).toBe(PropTypes.string);
    expect(propTypes.iconColor).toBe(PropTypes.string);
    expect(propTypes.iconPosition).toBeTruthy();
    expect(propTypes.maxLength).toBe(PropTypes.number);
    expect(propTypes.maxValue).toBe(PropTypes.number);
    expect(propTypes.minLength).toBe(PropTypes.number);
    expect(propTypes.minValue).toBe(PropTypes.number);
    expect(propTypes.onValueChange).toBe(PropTypes.func.isRequired);
    expect(propTypes.placeholder).toBe(PropTypes.string);
    expect(propTypes.size).toBeTruthy();
    expect(propTypes.type).toBe(PropTypes.string);
    expect(propTypes.validation).toBeTruthy();
    expect(propTypes.validationErrorIcon).toBe(PropTypes.string);
    expect(propTypes.validationSuccessIcon).toBe(PropTypes.string);
    expect(propTypes.value).toBe(PropTypes.string.isRequired);
  });
});

describe('defaultProps', () => {
  it('should set default values for optional props', () => {
    const lineInputWrapper = mount(<LineInput onValueChange={onValueChangeMock} value="abc" />);

    expect(lineInputWrapper.props().allowEmptyValue).toBe(false);
    expect(lineInputWrapper.props().className).toBe(undefined);
    expect(lineInputWrapper.props().countryCode).toBe('');
    expect(lineInputWrapper.props().disabled).toBe(false);
    expect(lineInputWrapper.props().errorText).toBe('');
    expect(lineInputWrapper.props().errorTextPosition).toBe('bottom');
    expect(lineInputWrapper.props().icon).toBe('');
    expect(lineInputWrapper.props().iconColor).toBe(undefined);
    expect(lineInputWrapper.props().iconPosition).toBe('left');
    expect(lineInputWrapper.props().maxLength).toBe(undefined);
    expect(lineInputWrapper.props().maxValue).toBe(undefined);
    expect(lineInputWrapper.props().minLength).toBe(undefined);
    expect(lineInputWrapper.props().minValue).toBe(undefined);
    expect(lineInputWrapper.props().placeholder).toBe('');
    expect(lineInputWrapper.props().size).toBe('small');
    expect(lineInputWrapper.props().type).toBe(undefined);
    expect(lineInputWrapper.props().validation).toBe(undefined);
    expect(lineInputWrapper.props().validationErrorIcon).toBe('');
    expect(lineInputWrapper.props().validationSuccessIcon).toBe('');
  });
});

describe('state', () => {
  it('should set initial state', () => {
    const lineInputWrapper = mount(<LineInput onValueChange={onValueChangeMock} value="abc" />);

    expect(lineInputWrapper.state('hasValidValue')).toBe(true);
    expect(lineInputWrapper.state('hasFocus')).toBe(false);
    expect(lineInputWrapper.state('hasBeenFocused')).toBe(false);
  });
});

describe('onInputBlur', () => {
  it('it should set hasValidValue in state to true, if no validation is provided', () => {
    const lineInputWrapper = mount(<LineInput onValueChange={onValueChangeMock} value="abc" />);
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('focus');
    inputWrapper.simulate('blur');

    expect(lineInputWrapper.state('hasValidValue')).toBe(true);
  });

  it('it should set hasValidValue in state to true, if validation is successful', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="number" value="123" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('focus');
    inputWrapper.simulate('blur');

    expect(lineInputWrapper.state('hasValidValue')).toBe(true);
  });

  it('it should not validate, if value is empty and allowEmptyValue is true', () => {
    const lineInputWrapper = mount(
      <LineInput allowEmptyValue onValueChange={onValueChangeMock} validation="number" value="" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('focus');
    inputWrapper.simulate('blur');

    expect(lineInputWrapper.state('hasValidValue')).toBe(true);
  });

  it('it should set hasValidValue in state to false and show error in input, if validation fails', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="number" value="abc" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('focus');
    inputWrapper.simulate('blur');
    lineInputWrapper.update();
    const newInputWrapper = lineInputWrapper.find('input');

    expect(lineInputWrapper.state('hasValidValue')).toBe(false);
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    expect(newInputWrapper.props().style.borderBottomColor).toBe('red');
  });

  it('it should show errorText, if validation fails', () => {
    const lineInputWrapper = mount(
      <LineInput errorText="errorText" onValueChange={onValueChangeMock} validation="number" value="abc" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('focus');
    inputWrapper.simulate('blur');
    lineInputWrapper.update();

    const labelWrapper = lineInputWrapper.find('div.label');
    expect(labelWrapper).toBeTruthy();
  });

  it('it should show errorText, if validation fails', () => {
    const lineInputWrapper = mount(
      <LineInput errorText="errorText" onValueChange={onValueChangeMock} validation="number" value="abc" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('focus');
    inputWrapper.simulate('blur');
    lineInputWrapper.update();

    const divWrapper = lineInputWrapper.find('div').first();
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    expect(divWrapper.props().style.flexDirection).toBe('column');
    const labelWrapper = lineInputWrapper.find('div.label');
    expect(labelWrapper).toBeTruthy();
  });
});

describe('onInputFocus', () => {
  it('should modify state correctly', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="number" value="abc" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    lineInputWrapper.setState({ hasValidValue: false });
    inputWrapper.simulate('focus');

    expect(lineInputWrapper.state('hasValidValue')).toBe(true);
    expect(lineInputWrapper.state('hasFocus')).toBe(true);
    expect(lineInputWrapper.state('hasBeenFocused')).toBe(true);
  });
});

describe('onInputChange', () => {
  it('should call onValueChange', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="number" value="abc" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('change', { target: { value: 'def' } });

    expect(onValueChangeMock).toHaveBeenCalledWith('def');
  });

  it('should format a credit card number', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="creditCardNumber" value="" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('change', { target: { value: '5345123455557777' } });

    expect(onValueChangeMock).toHaveBeenCalledWith('5345 1234 5555 7777');
  });

  it('should format a partial credit card number when only first group of digits', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="creditCardNumber" value="" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('change', { target: { value: '534' } });

    expect(onValueChangeMock).toHaveBeenCalledWith('534');
  });

  it('should format a partial credit card number when second group of digits', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="creditCardNumber" value="" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('change', { target: { value: '534567' } });

    expect(onValueChangeMock).toHaveBeenCalledWith('5345 67');
  });

  it('should format a partial credit card number when third group of digits', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="creditCardNumber" value="" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('change', { target: { value: '534567771' } });

    expect(onValueChangeMock).toHaveBeenCalledWith('5345 6777 1');
  });

  it('should format a partial credit card number when fourth group of digits', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="creditCardNumber" value="" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('change', { target: { value: '534567771111223' } });

    expect(onValueChangeMock).toHaveBeenCalledWith('5345 6777 1111 223');
  });

  it('should format a partial credit card expiry', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="creditCardExpiration" value="" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('change', { target: { value: '11' } });

    expect(onValueChangeMock).toHaveBeenCalledWith('11 / ');
  });

  it('should format a credit card verification code', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="creditCardVerificationCode" value="" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    inputWrapper.simulate('change', { target: { value: '34k' } });

    expect(onValueChangeMock).toHaveBeenCalledWith('34');
  });
});

describe('getInputType()', () => {
  it('should return type, if type prop is specified', () => {
    const lineInputWrapper = renderShallow(
      <LineInput onValueChange={onValueChangeMock} type="password" value="abc" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    expect(inputWrapper.props().type).toBe('password');
  });

  test.each([['emailAddress', 'email'], ['phoneNumber', 'tel'], ['url', 'url']])(
    'it should return HTML input type according to specified validation type when type is not defined',
    (validation, htmlInputType) => {
      const lineInputWrapper = renderShallow(
        <LineInput onValueChange={onValueChangeMock} validation={validation as ValidationType} value="abc" />
      );
      const inputWrapper = lineInputWrapper.find('input');

      expect(inputWrapper.props().type).toBe(htmlInputType);
    }
  );

  it('should return "text", if any other validation type is specified', () => {
    const lineInputWrapper = renderShallow(
      <LineInput onValueChange={onValueChangeMock} validation="usSSN" value="abc" />
    );
    const inputWrapper = lineInputWrapper.find('input');

    expect(inputWrapper.props().type).toBe('text');
  });
});

describe('getIconName()', () => {
  it('should not show validationErrorIcon nor validationSuccessIcon when input has not been focused', () => {
    const lineInputWrapper = mount(
      <LineInput
        onValueChange={onValueChangeMock}
        validation="number"
        validationErrorIcon="exclamation"
        validationSuccessIcon="check"
        value=""
      />
    );

    const iconWrapper = lineInputWrapper.find(Icon);
    expect(iconWrapper.props().name).toBe('');
  });

  test.each([['', 'exclamation', 'red'], ['xx', 'exclamation', 'red'], ['123', 'check', 'green']])(
    'it should show validationErrorIcon or validationSuccessIcon depending on entered value on input blur',
    (value, iconName, color) => {
      const lineInputWrapper = mount(
        <LineInput
          onValueChange={onValueChangeMock}
          validation="number"
          validationErrorIcon="exclamation"
          validationSuccessIcon="check"
          value={value}
        />
      );
      const inputWrapper = lineInputWrapper.find('input');

      inputWrapper.simulate('focus');
      inputWrapper.simulate('blur');
      lineInputWrapper.update();

      const iconWrapper = lineInputWrapper.find(Icon);
      expect(iconWrapper.props().name).toBe(iconName);
      expect(iconWrapper.props().color).toBe(color);
    }
  );

  it('should not show credit card icon when validation is "creditCardNumber" and input is empty', () => {
    const lineInputWrapper = mount(
      <LineInput onValueChange={onValueChangeMock} validation="creditCardNumber" value="" />
    );

    expect(lineInputWrapper.exists(Icon)).toBe(false);
  });

  test.each([
    ['4444', 'cc visa'],
    ['5444', 'cc mastercard'],
    ['3434', 'cc amex'],
    ['3600', 'cc diners club'],
    ['6011', 'cc discover'],
    ['3528', 'cc jcb'],
    ['9999', 'credit card']
  ])(
    'it should show credit card icon depending on entered value on input blur when validation is "creditCardNumber" ',
    (value, iconName) => {
      const lineInputWrapper = mount(
        <LineInput
          onValueChange={onValueChangeMock}
          validation="creditCardNumber"
          value={value}
        />
      );
      const inputWrapper = lineInputWrapper.find('input');

      inputWrapper.simulate('focus');
      inputWrapper.simulate('blur');
      lineInputWrapper.update();

      const iconWrapper = lineInputWrapper.find(Icon);
      expect(iconWrapper.props().name).toBe(iconName);
    }
  );

  it('should show icon specified in icon prop', () => {
    const lineInputWrapper = mount(
      <LineInput
        icon="search"
        onValueChange={onValueChangeMock}
        value="text"
      />
    );

    const iconWrapper = lineInputWrapper.find(Icon);
    expect(iconWrapper.props().name).toBe('search');
  })
});

describe('render()', () => {
  it('should render component correctly', () => {
    const lineInputWrapper = renderShallow(<LineInput onValueChange={onValueChangeMock} value="abc" />);

    expect(lineInputWrapper).toMatchSnapshot();
  });

  it('should render disabled component correctly', () => {
    const lineInputWrapper = renderShallow(
      <LineInput disabled onValueChange={onValueChangeMock} value="abc" />
    );

    expect(lineInputWrapper).toMatchSnapshot();
  });

  it('should render component with errorTextPosition "right" correctly', () => {
    const lineInputWrapper = renderShallow(
      <LineInput errorTextPosition="right" onValueChange={onValueChangeMock} value="abc" />
    );

    expect(lineInputWrapper).toMatchSnapshot();
  });

  it('should render validationErrorIcon correctly on right side', () => {
    const lineInputWrapper = renderShallow(
      <LineInput onValueChange={onValueChangeMock} validationErrorIcon="exclamation" value="abc" />
    );

    expect(lineInputWrapper).toMatchSnapshot();
  });

  it('should render validationSuccessIcon correctly on right side', () => {
    const lineInputWrapper = renderShallow(
      <LineInput onValueChange={onValueChangeMock} validationErrorIcon="exclamation" value="abc" />
    );

    expect(lineInputWrapper).toMatchSnapshot();
  });

  it('should render icon correctly depending on iconPosition and iconColor', () => {
    const lineInputWrapper = renderShallow(
      <LineInput icon="search" iconColor="blue" iconPosition="right" onValueChange={onValueChangeMock} value="abc" />
    );

    expect(lineInputWrapper).toMatchSnapshot();
  });

  it('should render component with additional HTML attributes correctly', () => {
    const lineInputWrapper = renderShallow(
      <LineInput
        className="test"
        id="test1"
        onValueChange={onValueChangeMock}
        style={{ visibility: 'hidden' }}
        value="abc"
      />
    );

    expect(lineInputWrapper).toMatchSnapshot();
  });
});
