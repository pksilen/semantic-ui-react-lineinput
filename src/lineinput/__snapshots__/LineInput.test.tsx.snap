// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() should render component correctly 1`] = `
<div
  style={
    Object {
      "display": "inline-flex",
      "flexDirection": "column",
    }
  }
>
  <div
    className="ui input small"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      style={
        Object {
          "borderLeft": "none",
          "borderRadius": 0,
          "borderRight": "none",
          "borderTop": "none",
          "paddingLeft": "0.5em",
          "paddingRight": "0.5em",
        }
      }
      type="text"
      value="abc"
    />
  </div>
</div>
`;

exports[`render() should render component with additional HTML attributes correctly 1`] = `
<div
  className="test"
  id="test1"
  style={
    Object {
      "display": "inline-flex",
      "flexDirection": "column",
    }
  }
>
  <div
    className="ui input small"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      style={
        Object {
          "borderLeft": "none",
          "borderRadius": 0,
          "borderRight": "none",
          "borderTop": "none",
          "paddingLeft": "0.5em",
          "paddingRight": "0.5em",
        }
      }
      type="text"
      value="abc"
    />
  </div>
</div>
`;

exports[`render() should render component with errorTextPosition "bottom" correctly 1`] = `
<div
  style={
    Object {
      "display": "inline-flex",
      "flexDirection": "column",
    }
  }
>
  <div
    className="ui input small"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      style={
        Object {
          "borderLeft": "none",
          "borderRadius": 0,
          "borderRight": "none",
          "borderTop": "none",
          "paddingLeft": "0.5em",
          "paddingRight": "0.5em",
        }
      }
      type="text"
      value="abc"
    />
  </div>
</div>
`;

exports[`render() should render disabled component correctly 1`] = `
<div
  style={
    Object {
      "display": "inline-flex",
      "flexDirection": "column",
    }
  }
>
  <div
    className="ui input small disabled"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      style={
        Object {
          "borderLeft": "none",
          "borderRadius": 0,
          "borderRight": "none",
          "borderTop": "none",
          "paddingLeft": "0.5em",
          "paddingRight": "0.5em",
        }
      }
      type="text"
      value="abc"
    />
  </div>
</div>
`;

exports[`render() should render icon correctly depending on iconPosition and iconColor 1`] = `
<div
  style={
    Object {
      "display": "inline-flex",
      "flexDirection": "column",
    }
  }
>
  <div
    className="ui input small right icon"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      style={
        Object {
          "borderLeft": "none",
          "borderRadius": 0,
          "borderRight": "none",
          "borderTop": "none",
          "paddingLeft": "0.5em",
          "paddingRight": "0.5em",
        }
      }
      type="text"
      value="abc"
    />
    <Icon
      as="i"
      color="blue"
      name="search"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`render() should render validationErrorIcon correctly on right side 1`] = `
<div
  style={
    Object {
      "display": "inline-flex",
      "flexDirection": "column",
    }
  }
>
  <div
    className="ui input small right icon"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      style={
        Object {
          "borderLeft": "none",
          "borderRadius": 0,
          "borderRight": "none",
          "borderTop": "none",
          "paddingLeft": "0.5em",
          "paddingRight": "0.5em",
        }
      }
      type="text"
      value="abc"
    />
    <Icon
      as="i"
      color="green"
      name=""
      style={
        Object {
          "opacity": 1,
        }
      }
    />
  </div>
</div>
`;

exports[`render() should render validationSuccessIcon correctly on right side 1`] = `
<div
  style={
    Object {
      "display": "inline-flex",
      "flexDirection": "column",
    }
  }
>
  <div
    className="ui input small right icon"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      style={
        Object {
          "borderLeft": "none",
          "borderRadius": 0,
          "borderRight": "none",
          "borderTop": "none",
          "paddingLeft": "0.5em",
          "paddingRight": "0.5em",
        }
      }
      type="text"
      value="abc"
    />
    <Icon
      as="i"
      color="green"
      name=""
      style={
        Object {
          "opacity": 1,
        }
      }
    />
  </div>
</div>
`;
